File: src/main/java/com/wesker926/leetcode/algorithms/p0519/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0519.s1;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;

/**
 * @author wesker.gh
 * @date 2021/11/27
 * @description 哈希空位映射
 */
public class Solution {

    private final Random random;

    private final int m, n;

    private final Map<Integer, Integer> map;

    private int size;

    /**
     * Your Solution object will be instantiated and called as such:
     * Solution obj = new Solution(m, n);
     * int[] param_1 = obj.flip();
     * obj.reset();
     */
    public Solution(int m, int n) {
        this.m = m;
        this.n = n;
        random = new Random();
        size = m * n;
        map = new HashMap<>();
    }

    public int[] flip() {
        int idx = random.nextInt(size);
        size--;

        int realIdx = map.getOrDefault(idx, idx);
        map.put(idx, map.getOrDefault(size, size));
        return new int[]{realIdx / n, realIdx % n};
    }

    public void reset() {
        size = m * n;
        map.clear();
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0519/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/11/27
// @description: 哈希空位映射
package s1

import "math/rand"

type Solution struct {
	m     int
	n     int
	size  int
	table map[int]int
}

func Constructor(m int, n int) Solution {
	return Solution{
		m:     m,
		n:     n,
		size:  m * n,
		table: make(map[int]int),
	}
}

func (this *Solution) Flip() []int {
	idx := rand.Intn(this.size)
	this.size--

	realIdx := idx
	if t, ok := this.table[idx]; ok {
		realIdx = t
	}

	replace := this.size
	if t, ok := this.table[this.size]; ok {
		replace = t
	}
	this.table[idx] = replace

	return []int{realIdx / this.n, realIdx % this.n}
}

func (this *Solution) Reset() {
	this.size = this.m * this.n
	this.table = make(map[int]int)
}

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(m, n);
 * param_1 := obj.Flip();
 * obj.Reset();
 */
